File: python/taflow/metrics/composite_profitability_consistency_index.py
"""Composite profitability and consistency metric."""
from __future__ import annotations
from typing import Any
from .._native.metrics import CompositeProfitabilityConsistencyIndex as _Native
from ._input import as_metric_series


class CompositeProfitabilityConsistencyIndex:
    """Multiply Profit Factor, decisive Win Rate, and Payoff Ratio.

    The independent oracle is QuantStats 0.0.81 ``cpc_index``. The complete
    TAFlow class name describes its formula and does not claim that “CPC” has a
    standardized historical expansion. Wins and losses are strict; breakevens
    remain in native length but are excluded from decisive win probability.
    Warm-up requires at least one win and one loss; otherwise the result is
    ``None``. Inputs are decimal returns or realized closed-trade P&L. Rust
    owns all allocation-free O(1) arithmetic and bulk work releases the GIL.
    """
    def __init__(self, nan_policy: str = "omit") -> None:
        """Initialize an empty configured metric."""
        self._state = _Native(nan_policy)

    def from_returns(self, returns: Any, *, column: str | None = None) -> "CompositeProfitabilityConsistencyIndex":
        """Append chronological returns observations and return this metric."""
        self._state.from_returns(as_metric_series(returns, column=column))
        return self

    def from_trades(self, trades: Any, *, column: str | None = None) -> "CompositeProfitabilityConsistencyIndex":
        """Append chronological trades observations and return this metric."""
        self._state.from_trades(as_metric_series(trades, column=column))
        return self

    def append(self,value:float)->"CompositeProfitabilityConsistencyIndex":
        """Append one observation and return this metric."""
        self._state.append(float(value));return self
    def extend(self,values:Any,*,column:str|None=None)->"CompositeProfitabilityConsistencyIndex":
        """Append observations and return this metric."""
        self._state.extend(as_metric_series(values,column=column));return self
    @property
    def value(self)->float|None:
        """Return the composite, or ``None`` during warm-up."""
        return self._state.value
    def compute(self)->float|None:
        """Return current scalar without replaying input."""
        return self._state.compute()
    def reset(self)->"CompositeProfitabilityConsistencyIndex":
        """Clear state, preserve domain, and return this metric."""
        self._state.reset();return self
    def __len__(self)->int:
        """Return valid observation count delegated to Rust."""
        return len(self._state)


__all__=["CompositeProfitabilityConsistencyIndex"]

File: crates/taflow-metrics/src/metrics/composite_profitability_consistency_index.rs
use crate::{
    primitives::GainLossState, MetricError, MetricInputKind, MetricInputState, MetricResult,
    NanPolicy,
};

/// Profit factor multiplied by decisive win rate and payoff ratio.
#[derive(Debug, Clone)]
pub struct CompositeProfitabilityConsistencyIndex {
    input: MetricInputState,
    gains: GainLossState,
}
impl CompositeProfitabilityConsistencyIndex {
    /// Construct an empty return or closed-trade state.
    pub fn new(nan_policy: NanPolicy) -> MetricResult<Self> {
        Ok(Self {
            input: MetricInputState::unbound(nan_policy),
            gains: GainLossState::new(),
        })
    }
    /// Append one observation and return the composite when defined.
    pub fn append(&mut self, value: f64) -> MetricResult<Option<f64>> {
        if let Some(observation) = self.input.append(value)? {
            self.gains.append(observation);
        }
        Ok(self.value())
    }
    /// Append observations through the same state.
    pub fn extend(&mut self, values: &[f64]) -> MetricResult<Option<f64>> {
        for &value in values {
            self.append(value)?;
        }
        Ok(self.value())
    }
    /// Return profit factor times decisive win rate times payoff ratio.
    pub fn value(&self) -> Option<f64> {
        let wins = self.gains.gain_count();
        let losses = self.gains.loss_count();
        if wins == 0 || losses == 0 {
            return None;
        }
        let gross_gain = self.gains.gross_gain();
        let gross_loss = self.gains.gross_loss().abs();
        let profit_factor = gross_gain / gross_loss;
        let win_rate = wins as f64 / (wins + losses) as f64;
        let payoff = (gross_gain / wins as f64) / (gross_loss / losses as f64);
        Some(profit_factor * win_rate * payoff)
    }
    /// Return current scalar without replay.
    pub fn compute(&self) -> Option<f64> {
        self.value()
    }
    /// Reset while preserving domain.
    pub fn reset(&mut self) {
        self.input.reset();
        self.gains.reset();
    }
    /// Return valid observation count, including breakevens.
    pub fn len(&self) -> usize {
        self.input.len()
    }
    /// Return whether no observations were processed.
    pub fn is_empty(&self) -> bool {
        self.input.is_empty()
    }
}

crate::impl_return_trade_metric_lifecycle!(CompositeProfitabilityConsistencyIndex);
